### DISCUSSION TITLE
Creating a static executable with cosmopolitan

### DISCUSSION THREAD

**Discussion by tecosaur** [janet-lang/janet#1363] [Q&A]
Hi, I'm trying to use Janet to create APE executables using the cosmopolitan C compiler + libc.

Looking at https://janet-lang.org/docs/jpm.html and https://github.com/janet-lang/jpm/blob/65e218ef55ab2fc601479708cdd9db35d10cfe05//jpm/config.janet#L255-L258 I've attempted to build following the instructions in https://github.com/jart/cosmopolitan#getting-started.

This is what I put in my `project.janet`

```lisp
(declare-executable
 :name "cli"
 :entry "cli.janet"
 :cc "x86_64-unknown-cosmo-cc"
 :cc-link "x86_64-unknown-cosmo-cc"
 :c++ "x86_64-unknown-cosmo-c++"
 :c++-link "x86_64-unknown-cosmo-c++")
```

If I run `jpm build --verbose` I can see that plain `cc` is used :(

Looking at https://github.com/janet-lang/jpm/blob/3d9d4c410fb64763620c3166c77f6340eae58ded/jpm/declare.janet#L205 I see that `declare-executable` doesn't actually accept `:cc` and friends.

It looks like I'd need to use `declare-native` for this, but I can see a way to easily get a binary entrypoint instead of a library.

If anyone could help me navigate this situation, it would be much appreciated.

**Comment by tionis** [comment]
It seems like jpm would need some changes to make this work. In theory it should just work when an custom `cc` is specified as in the background jpm just generates a c file to compile that imports the janet header.
The cosmopolitan version of Janet has no support for a number of features, so that should also be noted.

**Comment by strangepete** [comment]
I was able to get a cosmopolitan jpm hello world project to build, but it required building Janet with cosmo as well.

First, to build Janet I needed to remove the `-shared` library flag from the Makefile, and set the env as shown on the cosmo site:
```
export CC=x86_64-unknown-cosmo-cc
export CXX=x86_64-unknown-cosmo-c++
make -j
```

then in the project, set jpm libpath to your Janet `build/` directory, and set your cc/c++ flags:
```
jpm --cc=x86_64-unknown-cosmo-cc --c++=x86_64-unknown-cosmo-c++ --libpath=/home/pete/janet-cosmo/build build
```

finally, to export to windows:
```
x86_64-linux-cosmo-objcopy -SO binary build/cosmotest build/cosmotest.com
```

There will be quirks probably; for example when running the linux-compiled Janet REPL on Windows, it expects modules to have an ".so" extension and won't load dlls
